src/precice-aste-join.cpp

```diff
@@ -0,0 +1,320 @@
+    std::cerr << "Error: " << filename << " is not a valid output file." << std::endl;
+    std::exit(EXIT_FAILURE);
+  }
+}
+
+auto partitionwiseMerge(const std::string &prefix, size_t numparts) -> vtkSmartPointer<vtkUnstructuredGrid>
+{
+  auto             joinedMesh   = vtkSmartPointer<vtkUnstructuredGrid>::New();
+  auto             joinedPoints = vtkSmartPointer<vtkPoints>::New();
+  auto             joinedCells  = vtkSmartPointer<vtkCellArray>::New();
+  std::vector<int> joinedCellTypes;
+
+  std::vector<vtkSmartPointer<vtkDoubleArray>> joinedDataVec;
+  std::vector<std::string>                     joinedDatanames;
+
+  auto reader = vtkSmartPointer<vtkXMLUnstructuredGridReader>::New();
+  for (size_t i = 0; i < numparts; ++i) {
+    // Read mesh
+    auto partname = prefix + "_" + std::to_string(i) + ".vtu";
+    reader->SetFileName(partname.c_str());
+    reader->Update();
+    // Extract mesh
+    auto grid = reader->GetOutput();
+    // Cells
+    const auto offset   = joinedPoints->GetNumberOfPoints();
+    auto       numCells = grid->GetCells()->GetNumberOfCells();
+    joinedCellTypes.reserve(joinedCellTypes.size() + numCells);
+    auto cellIds = vtkSmartPointer<vtkIdList>::New();
+    for (int j = 0; j < numCells; ++j) {
+      cellIds->Reset();
+      std::for_each(grid->GetCell(j)->GetPointIds()->begin(), grid->GetCell(j)->GetPointIds()->end(), [&cellIds, &offset](auto &pointId) { cellIds->InsertNextId(pointId + offset); });
+      joinedCells->InsertNextCell(cellIds);
+      joinedCellTypes.push_back(grid->GetCellType(j));
+    }
+    // Points
+    auto points = grid->GetPoints();
+    joinedPoints->InsertPoints(joinedPoints->GetNumberOfPoints(), grid->GetNumberOfPoints(), 0, points);
+    //  Point Data
+    auto partPointData = grid->GetPointData();
+    auto numArrays     = partPointData->GetNumberOfArrays();
+    for (int j = 0; j < numArrays; ++j) {
+      auto partData = partPointData->GetArray(j);
+      auto name     = partData->GetName();
+      if (std::find(joinedDatanames.begin(), joinedDatanames.end(), name) == joinedDatanames.end()) {
+        joinedDatanames.emplace_back(name);
+        auto newJoinedData = vtkSmartPointer<vtkDoubleArray>::New();
+        newJoinedData->SetName(name);
+        newJoinedData->SetNumberOfComponents(partData->GetNumberOfComponents());
+        joinedDataVec.push_back(newJoinedData);
+      }
+      auto joinedData = joinedDataVec[j];
+      joinedData->InsertTuples(joinedData->GetNumberOfTuples(), partData->GetNumberOfTuples(), 0, partData);
+    }
+  }
+
+  joinedMesh->SetPoints(joinedPoints);
+  for (const auto &data : joinedDataVec) {
+    joinedMesh->GetPointData()->AddArray(data);
+  }
+  joinedMesh->SetCells(joinedCellTypes.data(), joinedCells);
+
+  return joinedMesh;
+}
+
+auto recoveryMerge(const std::string &prefix, std::size_t numparts, int size, const std::vector<int> &cellTypes, const std::vector<std::vector<int>> &cells) -> vtkSmartPointer<vtkUnstructuredGrid>
```

Comment:
Looks like there is quite a lot of duplication between the partitionwiseMerge and the recoveryMerge. I would expect that the one is a subroutine of the other. 
I would propose to first add tests and afterwards refactor here.